src/shims/foreign_items.rs

```diff
@@ -1032,6 +1032,29 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                 }
             }
 
+            // Used to implement the x86 `_mm{,256,512}_popcnt_epi{8,16,32,64}` and wasm
+            // `{i,u}8x16_popcnt` functions.
+            name if name.starts_with("llvm.ctpop.v") => {
+                let [op] = this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (op, op_len) = this.operand_to_simd(op)?;
+                let (dest, dest_len) = this.place_to_simd(dest)?;
+
+                assert_eq!(dest_len, op_len);
```

Comment:
You have this check in here and a lot of other intrinsics. I think this is a good thing to check, but an ICE is probably the wrong failure mode; this assertion can be tripped if someone declares their own LLVM intrinsics but does it wrong.

For example, this program ICEs:
```rust
#![feature(link_llvm_intrinsics, simd_ffi)]
#[allow(improper_ctypes)]
use std::arch::x86_64::*;

extern "C" {
    #[link_name = "llvm.x86.sse2.sqrt.pd"]
    fn oof(x: __m128d) -> __m256d;
}

fn main() {
    unsafe {
        oof(_mm_set_pd(0.0, 0.0));
    };
}
```

Reply:
https://github.com/rust-lang/miri/issues/3076

Reply:
We generally have the policy that ICEing when intrinsics are used incorrectly is fine. I think we should apply that to LLVM intrinsics as well. If https://github.com/rust-lang/rust/pull/116093 lands, people will get an appropriate warning when enabling the feature.